RFC: URL Service (TTL: expired)

streamich opened the pull request:
## Summary

This RFC contains proposal and implementation details for the *URL Service*, which will replace our existing *Short URL Service* and *URL Generator Service*.

GitHub issue: https://github.com/elastic/kibana/issues/87304

### [Rendered text](https://github.com/streamich/kibana/blob/rfc-0017-url-service/rfcs/text/0017_url_service.md)

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+   * @param state State object to provide to the URL generator.
+   * @param slug Optionally, provide a custom slug (ID) of the short URL.
+   */
+  create(
```
this should be discouraged (or should maybe not exist at all) as it promotes introducing non-explicit dependencies. Every plugin providing a generator should expose it on its public contract and every plugin depending on it should use/import it from there. The only exception might be general code that is not aware with what generators it might be working.

Take a look at embeddables for example. Lens embeddable is exposed on lens plugin and should be imported from there whenever you plan to embed lens. Dashboard however doens't know with what kind of embeddables it works with and is thus directly using the registry.

Without this in discover, where we know we rely on visualize plugin for the top chart we would not have it listed in required dependencies (as we would be relying on embeddable). user could disable the visualize plugin and with it break discover (which would expect that visualize embeddable is in the registry, but suddenly isnt

ppisljar replied:
cc @stacey-gammon we still follow this mindset right ?

Dosant replied:
@ppisljar, without this or a similar API, how would you link between, let's say, Discover <-> Dashboard ? I think cyclic deps are not allowed.

I think we need API like this. Just types should be explicit that this could throw or return null if URL generator is missing.

ppisljar replied:
good point, but introducing api like this we would still be introducing cyclic dependencies just in non-explicit way.

Dosant replied:
I think creating links between two apps is a valid use-case. If we can't express this relation via the core's plugin system (not sure optional plugins can be cyclic), then we will need this registry-like API

streamich replied:
I think you both have good points, that's why in the current and proposed API you can get hold of a URL generator in two ways: (1) immediately when you create it; (2) through a registry:

![image](https://user-images.githubusercontent.com/9773803/113265412-662ed780-92d4-11eb-9692-c101a2bcdf35.png)

1. In the first use case a plugin would itself expose its URL generator from its plugin contract, which Peter suggested.
2. In the second use case, if there are cyclical dependencies, or for whatever other reason, you can still fetch any URL generator from the registry, which Anton suggested.

mshustov replied:
> good point, but introducing api like this we would still be introducing cyclic dependencies just in non-explicit way.

Consider an example with `Discover <-> Dashboard`
```js
// Dashboard wants to create a link to Dashboard
share.url.shortUrls.create({
  slug: 'my-discover-query', // Optional human-readable URL slug.
  generator: 'DISCOVER_DEEP_LINKS',
  state: {
    indexPattern: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx',
  },
});
```
How can we be sure that the ` generator id` and `state` are valid?
It can be changed at any time by Discover, but Dashboard doesn't know about it. TypeScript won't help here: Basel doesn't allow circular imports between packages.

> Discover <-> Dashboard

Maybe it can wait for the Domain-based folder structure ?https://github.com/elastic/kibana/issues/71566#issuecomment-688887617
In this case, Discover and Dashboard plugins can share `folders` within [Analyse domain](https://github.com/elastic/kibana/issues/71566#issue-656209142)

Dosant replied:
> TypeScript won't help here: Basel doesn't allow circular imports between packages.

Should we consider runtime-based validation for this then? 
Btw, @streamich, I think this is needed anyway because we want to expose URL generation as an API, right? So should we look into configuring state shape like `@kbn/config-schema` does?

ppisljar replied:
the right solution is probably not a service like this but smaller plugins.

streamich replied:
I'm a bit confused, what is the outcome of this thread? How do you suggest to create a short URL?

This proposed way is due to us wanting to combine URL generators with short URL service, so that short URL service can do migrations automatically.

streamich replied:
TypeScript types can be imported from some third plugin/package.

> ... because we want to expose URL generation as an API, right?

Chatting with Peter today, it does not look like we want to expose URL generatrs as an API. Instead there will be a client-side redirect endpoint:

```
/app/goto/_redirect/DISCOVER_DEEP_LINKS?params={"indexPattern":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","highlightedField":"foo"}&paramsVersion=7.x
```

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+| __DELETE__            | `/api/short_url/<slug>`             | Endpoint for deleting an existing short URL.                                                                                             |
+| __POST__              | `/api/short_url/<slug>`             | Endpoint for updating information about an existing short URL.                                                                           |
+| __GET__ or __POST__   | `/api/short_url/<slug>/_resolve`    | Same as `GET /api/short_url/<slug>`, but this endpoint also increments the short URL access count counter and the last access timestamp. |
+| __GET__ or __POST__   | `/api/short_url/<slug>/_redirect`   | Redirects the user using HTTP 301 (permanent) redirect to the destination long URL.                                                      |
```
im confused by this GET or POST endpoints, could you split this up to make it clearer ?

streamich replied:
I've cleaned it up and updated the RFC, see the latest text.

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+
+#### The URL generator "redirect" endpoint
+
+The `/api/url/_redirect` endpoint will receive the necessary payload to
```
does this exist on client/server contracts as well ?

streamich replied:
Updated RFC to only have it on client. But there will still be a legacy endoint on server, from old short URL service.

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+
+| HTTP method           | Path                                | Description                                                                                                                              |
+|-----------------------|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| __GET__ or __POST__   | `/api/url/_redirect`                | Redirects the user using HTTP 302 (temporary) redirect. Receives URL generator ID and parameters in the POST body payload.               |
```
do we really want to 302, or should we use new platforms navigateWithState ? seems with the latter we will be able to navigate into apps that don't expose full state in the url without them providing special endpoint just for that case.

lukeelmers replied:
++ Yeah I think this is a key consideration, IMO the primary benefit of a redirect route is that we can use it to proxy state to apps which otherwise wouldn't support state in the url.

As long as apps are already handling state from `navigateToApp`, then they wouldn't need to do anything differently... the redirect route would just take the state and an app ID, and call `navigateWithState` on the client. (You lose the nice semantics of a proper 302 redirect however)

streamich replied:
My thinking was that currently Cloud has a lot of deep links into Dashboad and Discover and those links are sometimes breaking. They could use a redirect route with URL generator to redirect them to the correct link, and it would use the proper 302 redirect semantics as look mentions, and the browser/proxy would cache that redirect.

So external use like Cloud could use the server redirect endpoint because:

1. The current Cloud linking use case does not need to send location state to the app.
2. Nice 302 semantics.
3. Possibly more efficient, as to execute the `navigateToApp` we need to load Kibana client-side, but to redirect user or just construct a URL, when we don't need the location state, can be done without loading Kibana client-side.

But, of course, we can remove this endpoint and just use a browser-side equivalent.

streamich replied:
Discussing it with Peter today agreed to only have a client-side redirect for smaller API surface. Like so:

```
/app/goto/_redirect/DISCOVER_DEEP_LINKS?params={"indexPattern":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","highlightedField":"foo"}&paramsVersion=7.x
```

ppisljar commented on rfcs/text/0017_url_service.md:
````diff
@@ -0,0 +1,462 @@
+});
+```
+
+These short URL will be stored in saved objects of type `url` and will be
````
i think we should mention that we will not be storing any urls, we will be storing generatorId + state (which will allow us to do migrations)

streamich replied:
will do

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+| __GET__ or __POST__   | `/api/short_url/<slug>/_resolve`    | Same as `GET /api/short_url/<slug>`, but this endpoint also increments the short URL access count counter and the last access timestamp. |
+| __GET__ or __POST__   | `/api/short_url/<slug>/_redirect`   | Redirects the user using HTTP 301 (permanent) redirect to the destination long URL.                                                      |
+
+
```
we should probably also add a migrate endpoint. If we are promoting users to store state on their side (not creating short url,  using the _redirect endpoint) we need to provide a way for them to migrate the state they stored as well. i am not sure we want to do that as that complicates things. could we just always rely on short urls ?

ppisljar replied:
considering cloud use case, i think its a valid assumption that cloud would integrate in a way that it would create short urls for all the links they want to expose inside management apps (and thus not worry about state or migration of it)

streamich replied:
Agreed with Peter that instead of Cloud creating short URL saved objects, there will be `paramsVersion` parameter in the redirect endpoint which will on-the-fly migrate the locator params if needed.

```
/app/goto/_redirect/DISCOVER_DEEP_LINKS?params={"indexPattern":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","highlightedField":"foo"}&paramsVersion=7.x
```

ppisljar commented on rfcs/text/0017_url_service.md:
````diff
@@ -0,0 +1,462 @@
+```ts
+const shortUrl = await share.url.shortUrls.create({
+  generator: 'DISCOVER_DEEP_LINKS',
+  state: {
````
how does the consumer know about this state ?|
- are we hoping to make this a very small set of settings the user could set ?
- or is the plan to have full state in here ?

maybe we need to separate the two ? if discover app wants to generate the short url, it has full knowledge of state. In this case i am not sure we should call this url generators at all. 
- discover could just use short url service. the short url saved object should contain appId and appState, and the endpoint should redirect with core.app.navigateWithState(appId, appState)

when someone external however wants to navigate to discover, they don't know the full state, but ight know some information. In this case the generators make sense. Even here they probably shouldn't be called url generators, but state generators. i want to navigate to discover and select an index pattern .....
- discoverPlugin.navigate({ indexPattern }) maybe just a method on discover plugin for that ?
- or the state generator, which takes in this simple object and returns full state i can then use with core.app.navigateTo ?

ppisljar replied:
i am not sure i fully follow ? if we would to make filters part of that public API .... we need to make filter structure public, or provide some other public API to generate this filters (which we already have)

streamich replied:
Agreed with Peter to rename "URL generators" to "locators". Also see the latest proposed API for creating short URLs.

```ts
const shortUrl = await plugins.discover.locator.createShortUrl(
  {
    indexPattern: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
    highlightedField: 'foo',
  }
  'human-readable-slug',
});
```

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+   destination page might have other state which affects the display of the page
+   but is not present in the URL. Once the short URL is used to navigate to that
+   page, any state that is kept only in memory is lost.
+   1. __Could do:__ We will not support storing additional, non-URL state in
```
if url generators would not be returning urls (so, they probably shouldn't be called url generators as mentioned in one of my ther comments) we can avoid this.

streamich replied:
Yes, this will be in the initial phase. I've first marked it as "Could do", but then actually expanded the API and talking with @ppisljar it will be in the initial phase.

joshdover commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+ */
+interface IUrlGeneratorClient {
+  register<S>(urlGenerator: UrlGeneratorDefinition<S>): UrlGenerator;
+  get<S>(id: string): UrlGenerator<S>;
```
Shouldn't this return `UrlGenerator<S> | undefined` if the generator requested is not registered (in case the plugin is disabled?)

streamich replied:
Could be either way. For example, it could throw if the requested generator is not available.

joshdover replied:
> it could throw if the requested generator is not available.

Yeah but that doesn't force the consumer to actually handle the exception. They have to read the docs to know that throwing is even possible.

lukeelmers reviewed (commented):
This might be a long comment, so apologies in advance... I've been thinking about this problem for over a year now, so I have a lot of thoughts 🙂 🙈   

But suffice it to say I'm glad to see us finally tackling this problem.

---

The use case for short urls is clear, but IMHO the need for a shared generator service is still ambiguous (vs putting everything in plugin contracts).

One thing that I think would be useful in this RFC is a breakdown of when we would recommend to use:
- URL generator from a plugin contract, vs
- URL generator via `generatorClient.get`, vs
- `core.app.navigateToApp`

I think that would help to answer the concerns raised by @ppisljar and @joshdover around creating implicit dependencies.

For the case of apps with circular dependencies, what's the benefit of using URL generators vs using `navigateToApp`?
- `generatorClient.get` requires you to pass a known ID, which functionally is no different than the app ID you pass to `navigateToApp`
- URL generator lets you pass state, so does `navigateToApp`
- URL generator handles migrations which the app registers, but each app could be handling migrations behind the scenes for any state that's passed via `navigateToApp`
- In either case you won't have type safety anyway due to the circular dep

The only downside I could think of would be a case where there's a circular dependency, and one app is storing the state somewhere and needs a way to migrate it. In that case the app that they are navigating to would be responsible for migrating the state on-the-fly as there wouldn't be a way for them to explicitly call `migrate`

---

I guess what I'm getting hung up on is that we keep talking about URLs, but in my mind the problem we are really trying to solve is around sharing state and has nothing to do with URLs.

I'm probably oversimplifying, but an alternative approach I have in my mind would be:
- Update the short url service as described here
  - This needs to be cleaned up anyway
- Implement a redirect route which takes an app ID & state, and hands it off to each app via `navigateToApp`
  - This solves for external users (i.e. Cloud)
- Provide a standardized interface which apps can use to add navigate / migrate functionality to their plugin contracts
  - So everyone just goes through plugin contracts on both client & server
  - If that creates a circular dependency, they use `navigateToApp` (client), or the redirect route (server) as a workaround (or maybe we implement `navigateToApp` on the server in core)
- Each app exports an interface of whatever state they consider "public"
  - This becomes the source of truth for what state is supported during navigation
  - They would also need to handle incoming state from `navigateToApp` and automatically migrate it on-the-fly

The concept of "URL state" is then entirely out of the question: If apps want to put state in URLs, that becomes an implementation detail.

Again, I'm probably oversimplifying, so please help me poke holes in this as I'm sure there are some use cases I haven't considered. ❤️

ppisljar commented:
@lukeelmers, i agree with your first point and i think even for migration purpose URL generators service doesn't bring any benefit. All the app needs to do is implement PeristableStateService as part of their contract. (for example discover app could provide `discover.navigateState: PersistableStateService` which would provide a way to migrate, collect telemetry and inject/extract variables for the navigateState)

I agree with proposed plan as well.

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+One alternative is still do the short URL improvements outlined above. But
+reconsider URL generators:
+
+- Do we need URL generators at all?
```
I think we could avoid circular dependencies by creating smaller plugins

streamich replied:
@joshdover It is a good question, I was also asking this a few times since the URL generators were introduced.

In general, I believe (as most devs here) that URL generators should be used from plugin contracts (which is done already now), maybe even through some application service API.

But there is also a registry of all URL generators, I don't now how good any of these reasons are, but here is what I collected:

1. As you mention, to avoid circular dependencies. Like Anton mentioned, consider Discover app wants to build a URL to Dashboard and Dashboard app wants to build a URL to Discover.
2. Help with cases when some service needs to persist the URL generators state, and it uses multiple URL generators. When it needs to use it, it then loads the generator ID and generator state (which is an opaque object from its perspective) and uses it: to navigate or to do a migration. Otherwise those services need to implement a registry of URL generators themselves internally.
3. We wanted to combine URL generators with short URL service, where you create a short URL using generator ID and its state. This way the short URL service handles automatically the state migrations (currently URL generators provide a way for you to migrate the state you feed into the generator).
4. For external, for example Cloud, to be able to use a single endpoint for any URL generator, we need a central registry of those.

streamich commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+|-----------------------|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| __ANY__               | `/goto/<slug>`                      | Legacy endpoint for redirecting short URLs. We will keep this endpoint, it will work the same as the `/go/<slug>` endpoint.              |
+| __GET__               | `/api/short_url/<slug>`             | The new `GET /api/short_url/<slug>` endpoint will return a superset of the payload that the legacy endpoint now returns.                 |
+| __POST__              | `/api/shorten_url`                  | The legacy endpoints for creating short URLs. We will remove it or deprecate this endpoint and maintain it until 8.0 major release.      |
```
Unfortunately, it is documented, here: https://www.elastic.co/guide/en/kibana/current/url-shortening-api.html

In Kibana proper it is used by `share` plugin to generate short URLs for Discover, Dashboard, Visualize apps.

When would you suggest removing it, do we have guidelines on deprecation/breaking changes?

joshdover replied:
We don't right now, but I would recommend leaving it as-is but deprecating it as soon as the new API is available. We'll need to give users a more gradual migration path than removing in 8.0.

streamich commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+| __GET__               | `/api/short_url/<slug>`             | Endpoint for retrieving information about an existing short URL.                                                                         |
+| __DELETE__            | `/api/short_url/<slug>`             | Endpoint for deleting an existing short URL.                                                                                             |
+| __POST__              | `/api/short_url/<slug>`             | Endpoint for updating information about an existing short URL.                                                                           |
+| __GET__ or __POST__   | `/api/short_url/<slug>/_resolve`    | Same as `GET /api/short_url/<slug>`, but this endpoint also increments the short URL access count counter and the last access timestamp. |
```
The motivation is to provide an API on the client-side to resolve short URL and increment the "hit" counter in one HTTP request. I think it is ok to use `GET` request here as it does not really do modifications, just increments a counter. But we can also make it `POST`, or reconsider if it is needed at all. The main idea was to provide the same `IUrlService` client interface on, both, client and server.

joshdover replied:
Only risk is this being used in an anchor tag's `href` and getting prefetched by the browser.

streamich commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,462 @@
+
+1. URL generators are available only on the client-side. There is no way to use
+   them together with short URLs.
+   1. __Will do:__ We will implement URL generators also on the server-side
```
Idea is to change the current RFC so that URL generators return `KibanaLocation`, where:

```ts
interface KibanaLocation {
  app: string; // app ID
  path: string;  // path in app
  state: object; // object to pass using navigateWithState()
}
```

so basically it is a relative URL, and we solve separately how it is converted to an absolute URL.

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -55,6 +55,12 @@ and in addition will implement the following improvements:
+# Terminology
+
+In the proposed new service we introduce "locators". This is mostly a change
+in language, we are renaming "URL generators" to "locators".
```
```suggestion
in language, we are renaming "URL generators" to "locators" as they are no longer generating any urls.
```

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -258,58 +349,80 @@ browser, wherever they choose to use the URL Service:
+  /** Creates a new short URL saved object using this locator. */
+  createShortUrl(params: P, slug?: string): Promise<ShortUrl>;
+  /** Returns a relative URL to the client-side redirect endpoint using this locator. */
+  getRedirectPath(params: P): string;
```
i would mark this one as internal and not intended to be used by external consumers and mention short urls are the preferred way to deep link from external sources

streamich replied:
Will do.

streamich commented:
I believe I've addressed all comments in this PR and updated the RFC as discussed with Peter with our latest thinking on this service. (I've marked all comments as "Resolved" to help me know I've seen or commented on it, it does not necessarily mean that the comment at hand is entirely "resolved" in the scope of this PR.)

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -324,32 +437,59 @@ interface IShortUrlClient {
 
-Note, that in the new service to generate a short URL the developer will have to
-use a URL generator.
+Note, that in this new service to create a short URL the developer will have to
```
should we even have this ? seems we can achieve everything thru locators:
`plugins.discover.locator.createShortUrl(...)` if we know the target locator at compile time or
`plugins.share.locators.get(locatorId).createShortUrl(state)` when we don't know locator at compile time

ppisljar replied:
(the create method)

streamich replied:
I've added it here as it felt alien to have a CRUD API without the "create" method. But it is not strictly necessary, removed in the RFC.

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,587 @@
+We have use cases were a deep link to some Kibana app is sent out, for example,
+through e-mail or as a Slack message.
+
+In this example, lets consider some plugin gets hold of the Discover locator
```
i think we should rather recommend generating a short url here: `plugins.discover.locator.getShortUrl(...)` as it has the benefit of us being able to migrate those.

streamich replied:
I don't think I like this idea of generating a new short URL (which is a new saved object) every time somebody simply needs a deep link into some Kibana app.

1. It is slow.
2. It creates a new saved object.
3. It will not work for read-only anonymous users.
4.

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,587 @@
+endpoint will receive the locator ID and locator params, it will use those to
+find the locator and execute `locator.navigate(params)` method.
+
+The optional `paramsVersion` parameter will be used to specify the version of the
```
should this be optional or mandatory? hopefully noone is generating this links by hand but uses `locator.getLocationPath() ` which should always include kibana version.

streamich replied:
Yeah, good point, maybe we can make it required, and the endpoint would return an error if it is not provided.

streamich replied:
I've made it required.

streamich commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,580 @@
+const locator = plugins.share.locators.create({
+  id: 'DISCOVER_DEEP_LINKS',
+  getLocation: ({
+    indexPattern,
```
I haven't thought about it, but, yes, it looks like we should.

Dosant replied:
Great, let's mention it somewhere?

streamich replied:
Yes, let's mention it, but let me think about it a bit. Also @ppisljar what do you think?

ppisljar replied:
i think _redirect endpoint brings a lot of problems and yes, this is definitely one of them.

streamich replied:
After thinking about this and discussing with @ppisljar we came up with the following:

- The redirect endpoint (if we will support it) will exposed only on the client side, not server.
- On the client apps already need to validate parameters they receive from the URL. And our thinking is that the parameters the apps receive from the location state should follow the same logic, apps should validate those the same as they validate URL parameters. Which means we don't need any extra validation here.

streamich commented on rfcs/text/0017_url_service.md:
````diff
@@ -0,0 +1,580 @@
+Discover application, they use this exported locator.
+
+```ts
+plugins.discover.locator.navigate({
````
In you case, you will be able to use the locator from the `share` plugin.

```ts
plugins.share.locators.get(locatorId);
```

Dosant replied:
Thanks! sounds good. 
I think I missed it in the latest version of the RFC

streamich replied:
We don't mention it explicitly, as we don't encourage to use it. There is this one line in the interface:

https://github.com/elastic/kibana/pull/95416/files#diff-586be16093047bf474d38759844a5379a1c55e4ff2dc1b492d31991bebc04453R373

ppisljar commented on rfcs/text/0017_url_service.md:
```diff
@@ -0,0 +1,585 @@
+provide two main functionalities:
+
+1. It will provide a facility to create locators.
+1. It will also be a registry of locators, every newly created locator is
```
lets mention when this registry should/shouldnt be used:

`the registry should never be used when locator ID is known at the compile time, but is resolved only for usecases where we only know ID of locator at the runtime.`

streamich replied:
Added you comment.

elasticmachine commented:
Pinging @elastic/kibana-app-services (Team:AppServices)

ppisljar reviewed (approved):
LGTM

Dosant reviewed (approved):
LGTM,
After the latest review had a tranch of questions which we discussed with @streamich offline. 
I am leaving key takeaways here: 

1. Current RFC version briefly mentions `paramsVersion=7.x` and state migrations. From the discussion, the plan is that Locator implements PersistableState interface and Locator params persisted in the stored saved objects should be migrated when updating Kibana. The same logic is used when migrating `_redirect` URLs on the fly. 
2. Was curious how do we teach teams deep-linking to Kibana to use endpoints as `/app/goto/_redirect/DISCOVER_DEEP_LINKS?params={"indexPattern":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","highlightedField":"foo"}&paramsVersion=7.x` instead of copying URLs from the browser window. Where do they learn about these URLs and how do they know about available params? Looks like each `Locator` should be documented separately and documentation for all of them should be centralized somewhere. Maybe some dev tool/playground will be needed for exploration and validation. 
3. In the current design locators/endpoints do not validate params. Apps validate incoming states when navigated to them. 
4. In the current design we do not worry about `params` length. If URL is too long -> use a short URL. 

--- 

We also reviewed this RFC from `search-sessions` perspective and looks like it would work. Comparing to URL generators this RFC unlocks navigation with history state (can support with this Lens editor and Canvas). 
We will need to add server-side migrations to migrate app state inside the search-session saved objects using locators.

streamich commented on rfcs/text/0017_url_service.md:
````diff
@@ -0,0 +1,600 @@
+```
+
+This redirect client-side endpoint would find the Discover locator and and
+execute the `.navigate()` method on it.
````
If I remember correctly, `core.navigateToApp` allows specifying if it is "push" or "replace".

Dosant replied:
Perfect!
